erts/emulator/beam/jit/arm/instr_bs.cpp

```diff
@@ -1028,18 +1028,192 @@ void BeamModuleAssembler::emit_i_bs_put_utf8(const ArgLabel &Fail,
     }
 }
 
+/*
+ * ARG1 = pointer to match state
+ * ARG2 = 4 bytes read from the binary in big-endian order
+ * ARG3 = position in binary in bits
+ *
+ * On successful return, the extracted code point is in ARG1 and the
+ * position in the match state has been updated. On failure,
+ * ARG1 is THE_NON_VALUE.
+ */
+void BeamGlobalAssembler::emit_bs_get_utf8_shared() {
+    const int position_offset = offsetof(ErlBinMatchBuffer, offset);
+
+    const arm::Gp match_state = ARG1;
+    const arm::Gp bitdata = ARG2;
+    const arm::Gp bin_position = ARG3;
+
+    const arm::Gp byte_count = ARG4;
+
+    const arm::Gp shift = TMP4;
+    const arm::Gp control_mask = TMP5;
+    const arm::Gp error_mask = TMP6;
+
+    /* UTF-8 has the following layout, where 'x' are data bits:
+     *
+     * 1 byte:  0xxxxxxx (not handled by this path)
+     * 2 bytes: 110xxxxx, 10xxxxxx
+     * 3 bytes: 1110xxxx, 10xxxxxx 10xxxxxx
+     * 4 bytes: 11110xxx, 10xxxxxx 10xxxxxx 10xxxxxx
+     *
+     * Note that the number of leading bits is equal to the number of bytes,
+     * which makes it very easy to create masks for extraction and error
+     * checking. */
+
+    /* Calculate the number of bytes. */
+    a.cls(byte_count, bitdata);
+    a.add(byte_count, byte_count, imm(1));
+
+    /* Get rid of the prefix bits. */
+    a.lsl(bitdata, bitdata, byte_count);
+    a.lsr(bitdata, bitdata, byte_count);
+
+    /* Calculate the bit shift now before we start to corrupt the
+     * byte_count. */
+    mov_imm(shift, 64);
+    a.sub(shift, shift, byte_count, arm::lsl(3));
+
+    /* Shift down the value to the least significant part of the word. */
+    a.lsr(bitdata, bitdata, shift);
+
+    /* Matches the '10xxxxxx' components, leaving the header byte alone. */
+    mov_imm(error_mask, 0x00808080ull << 32);
+    a.lsr(error_mask, error_mask, shift);
+
+    /* Construct the control mask '0x00C0C0C0' (already shifted). */
+    a.orr(control_mask, error_mask, error_mask, arm::lsr(1));
+
+    /* Assert that the header bits of each '10xxxxxx' component are correct,
+     * signaling errors by trashing the byte count with an illegal
+     * value (0). */
+    a.and_(TMP3, bitdata, control_mask);
+    a.cmp(TMP3, error_mask);
+    a.csel(byte_count, byte_count, ZERO, imm(arm::CondCode::kEQ));
+
+    a.ubfx(TMP1, bitdata, imm(8), imm(6));
+    a.ubfx(TMP2, bitdata, imm(16), imm(6));
+    a.ubfx(TMP3, bitdata, imm(24), imm(3));
+    a.ubfx(bitdata, bitdata, imm(0), imm(6));
+
+    a.orr(bitdata, bitdata, TMP1, arm::lsl(6));
+    a.orr(bitdata, bitdata, TMP2, arm::lsl(12));
+    a.orr(bitdata, bitdata, TMP3, arm::lsl(18));
+
+    /* Check for too large code point. */
+    cmp(bitdata, 0x10FFFF);
```

Comment:
We ought to be able to get rid of another `CSEL` by making this a (register-variant) `CCMP` too.

Reply:
Done.